### DISCUSSION TITLE
Raw terminal mode

### DISCUSSION THREAD

**Discussion by lelvisl** [vmware/govmomi#2802] [Community Help] [answered]
Is there a possibility in the library to implement raw terminal mode, like `kubectl exec bash`?

**Comment by dougm** [comment, marked as answer]
@hickeng implemented `docker exec` in https://github.com/vmware/vic , but there was quite a bit more involved than just govmomi.

Out of the box, govmomi only supports running a single command, as used in [govc guest.run](https://github.com/vmware/govmomi/blob/master/govc/USAGE.md#guestrun)

**Comment by hickeng** [comment]
Yeah, there's *a lot* more to it in terms of pty handling and other facets in the guest if wanting to be able to do something like use guest process manager to launch `bash` and then attach to it.

However the [lib/vspc](https://github.com/vmware/vic/tree/master/lib/vspc) package in VIC implements a client for the [Virtual Serial Port Concentrator](https://docs.vmware.com/en/VMware-vSphere/7.0/com.vmware.vsphere.vm_admin.doc/GUID-A779C84D-4324-4D32-AF43-BBD962ABC9E3.html) and the [lib/portlayer/attach](https://github.com/vmware/vic/tree/master/lib/portlayer/attach) package has the logic to add an appropriately configured serial port to a VM for this purpose.
We also needed to add [an handshake mechanism](https://github.com/vmware/vic/tree/master/pkg/serial) to clear some leading garbage characters - never managed to track down whether these were coming from Linux or somewhere in the vSphere stack.

That could be turned into a general serial console mechanism, but would still depends on the guest being configured for such, and the serial port configured ahead of time as I don't recall it being possible to hot-add (could be wrong).
